### PR TITLE
Remove violation detector from SecurityAnalysis API

### PR DESCRIPTION
**Pull request by flo-dup**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines

**Does this PR already have an issue describing the problem?**
No

**What kind of change does this PR introduce?**
Refactor

**What is the current behavior?**
LimitViolationDetector in SecurityAnalysis API, not in line wit

**What is the new behavior (if this is a feature change)?**
- `LimitViolationDetector` removed from `SecurityAnalysis`, `SecurityAnalysisProvider` and `SecurityAnalysisInput`
- interface `LimitViolationDetector` moved to dedicated package `com.powsybl.security.detectors`
- `LimitViolationDetector` parameter in `DefaultSecurityAnalysis` now optional (nullable)

**Does this PR introduce a breaking change or deprecate an API?**
- [x] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**
- [x] The *Breaking Change* or *Deprecated* label has been added
- [x] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**
Starting from this release the `LimitViolationDetector` is removed from security analysis API. The default implementations remains available, and the `DefaultSecurityAnalysis` default implementation still handles a `LimitViolationDetector` parameter through its constructor, although it is now becoming nullable. The `DefaultSecurityAnalysisProvider` now builds a `DefaultSecurityAnalysis` with a null `LimitViolationDetector`.

If you want to still use the `LimitViolationDetector` you need to pass one to your security analysis (or to the `DefaultSecurityAnalysis`). Hence, you cannot use directly the `DefaultSecurityAnalysisProvider` anymore. It is thus not possible to call `SecurityAnalysis.find()` to create the `SecurityAnalysis.Runner` anymore. Instead, you should declare your own `SecurityAnalysisProvider` (which may or not extends `DefaultSecurityAnalysisProvider`) which will create a `DefaultSecurityAnalysis` using the constructor taking a `LimitViolationDetector`. As a consequence, the provider won't be retrieved dynamically from the `default-impl-name` property of the configuration file.

### CODE REVIEW

File: security-analysis/security-analysis-default/src/main/java/com/powsybl/security/impl/DefaultSecurityAnalysis.java [hunk 1]
```diff
@@ -284,4 +292,22 @@ private ThreeWindingsTransformerResult createThreeWindingsTransformerResult(Thre
...
+        } else {
+            LimitViolationDetection.checkAll(network, EnumSet.allOf(LoadingLimitType.class), 1., consumer);
+        }
+    }
+
+    protected void checkPostContingencyViolations(Contingency contingency, Network network, Consumer<LimitViolation> consumer) {
+        if (violationDetector != null) {
+            violationDetector.checkAll(contingency, network, consumer);
+        } else {
+            // TODO For now, contingencies are ignored in the default security analysis
```
Comment (flo-dup) [inline security-analysis/security-analysis-default/src/main/java/com/powsybl/security/impl/DefaultSecurityAnalysis.java:307]: Maybe remove as already discussed the `TODO For now` and `This should change to fully support LimitReductions`, and add something like `This blocks a full support of LimitReductions`. Indeed we won't probably change this behaviour in the future.

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/detectors/LimitViolationDetector.java [hunk 2]
```diff
@@ -483,7 +476,7 @@ default void checkPermanentLimit(ThreeWindingsTransformer transformer, ThreeSide
     /**
      * Generic implementation for temporary limit checks
```
[the remark this answers is not in the document]
Comment (olperr1) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/detectors/LimitViolationDetector.java:477]: You are right.
I changed the returned value (`true` when no overload is detected) and updated the javadoc.

File: security-analysis/security-analysis-default/src/main/java/com/powsybl/security/impl/DefaultSecurityAnalysis.java (hunk 1 above)
Comment (olperr1) [reply security-analysis/security-analysis-default/src/main/java/com/powsybl/security/impl/DefaultSecurityAnalysis.java]: I removed the `TODO` and added a clearer comment.

**Review by flo-dup — COMMENTED** [review]
The PR is now ok for me!

**Review by olperr1 — APPROVED** [review]
PR OK with @flo-dup (see previous comment)

**Pull request merged**